File: examples/tito_example.go
// Together In Together Out - Go example.
//
// Run: go run examples/tito_example.go

package main

import (
	"fmt"
	"strings"
)

type group struct {
	id      string
	members []string
	ready   map[string]bool
}

// titoQueue is a minimal strict-order (head-of-line FIFO) TITO queue.
type titoQueue struct {
	order   []*group
	ids     map[string]bool
	groupOf map[string]*group
}

func newTitoQueue() *titoQueue {
	return &titoQueue{
		ids:     make(map[string]bool),
		groupOf: make(map[string]*group),
	}
}

func (q *titoQueue) admit(id string, members []string) error {
	if len(members) == 0 {
		return fmt.Errorf("a group must contain at least one member")
	}
	if q.ids[id] {
		return fmt.Errorf("duplicate group id: %s", id)
	}
	g := &group{id: id, members: append([]string(nil), members...), ready: make(map[string]bool)}
	for _, member := range members {
		if _, taken := q.groupOf[member]; taken {
			return fmt.Errorf("duplicate member: %s", member)
		}
		if _, seen := g.ready[member]; seen {
			return fmt.Errorf("duplicate member: %s", member)
		}
		g.ready[member] = false
	}
	for _, member := range members {
		q.groupOf[member] = g
	}
	q.ids[id] = true
	q.order = append(q.order, g)
	return nil
}

func (q *titoQueue) markReady(member string) error {
	g, ok := q.groupOf[member]
	if !ok {
		return fmt.Errorf("unknown member: %s", member)
	}
	g.ready[member] = true
	return nil
}

// release returns the head group, or nil while it is not fully ready.
func (q *titoQueue) release() *group {
	if len(q.order) == 0 {
		return nil
	}
	g := q.order[0]
	for _, member := range g.members {
		if !g.ready[member] {
			return nil
		}
	}
	q.order = q.order[1:]
	delete(q.ids, g.id)
	for _, member := range g.members {
		delete(q.groupOf, member)
	}
	return g
}

func main() {
	queue := newTitoQueue()

	if err := queue.admit("party-1", []string{"ann", "bob"}); err != nil {
		panic(err)
	}
	fmt.Println("admit party-1: ann, bob")
	if err := queue.admit("party-2", []string{"cy"}); err != nil {
		panic(err)
	}
	fmt.Println("admit party-2: cy")

	for _, member := range []string{"ann", "bob", "cy"} {
		if err := queue.markReady(member); err != nil {
			panic(err)
		}
		fmt.Printf("ready %s\n", member)
		if released := queue.release(); released == nil {
			fmt.Println("release -> waiting")
		} else {
			fmt.Printf("release -> %s [%s]\n", released.id, strings.Join(released.members, " "))
		}
	}
}

File: examples/tito_example.kt
// Together In Together Out - Kotlin example.
//
// Run: kotlinc examples/tito_example.kt -include-runtime -d /tmp/tito.jar
//      && java -jar /tmp/tito.jar

/** A group released as a whole. */
data class Released(val groupId: String, val members: List<String>)

/** Minimal strict-order (head-of-line FIFO) TITO queue. */
class TitoQueue {
    private val order = ArrayDeque<String>()
    private val groups = HashMap<String, LinkedHashMap<String, Boolean>>()
    private val groupOf = HashMap<String, String>()

    fun admit(groupId: String, members: List<String>) {
        require(members.isNotEmpty()) { "a group must contain at least one member" }
        require(!groups.containsKey(groupId)) { "duplicate group id: $groupId" }
        val ready = LinkedHashMap<String, Boolean>()
        for (member in members) {
            require(!groupOf.containsKey(member) && !ready.containsKey(member)) {
                "duplicate member: $member"
            }
            ready[member] = false
        }
        for (member in members) {
            groupOf[member] = groupId
        }
        groups[groupId] = ready
        order.addLast(groupId)
    }

    fun markReady(member: String) {
        val groupId = requireNotNull(groupOf[member]) { "unknown member: $member" }
        groups.getValue(groupId)[member] = true
    }

    /** Releases the head group, or null while it is not fully ready. */
    fun release(): Released? {
        val groupId = order.firstOrNull() ?: return null
        val ready = groups.getValue(groupId)
        if (ready.values.any { !it }) {
            return null
        }
        order.removeFirst()
        groups.remove(groupId)
        val members = ready.keys.toList()
        for (member in members) {
            groupOf.remove(member)
        }
        return Released(groupId, members)
    }
}

fun main() {
    val queue = TitoQueue()

    queue.admit("party-1", listOf("ann", "bob"))
    println("admit party-1: ann, bob")
    queue.admit("party-2", listOf("cy"))
    println("admit party-2: cy")

    for (member in listOf("ann", "bob", "cy")) {
        queue.markReady(member)
        println("ready $member")
        val released = queue.release()
        if (released == null) {
            println("release -> waiting")
        } else {
            println("release -> ${released.groupId} [${released.members.joinToString(" ")}]")
        }
    }
}
